README
Resque version 0.01
===================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Test

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2011 by Allen Fair

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.3 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Resque.

0.01  Fri Aug 26 18:34:40 2011
	- original version; created by h2xs 1.23 with options
		-b 5.6.0 -AX Resque



lib/Resque.pm
package Resque;

use 5.006000;
use strict;
use warnings;

use Redis;
use Resque::Worker;
use JSON::XS;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Resque ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

sub new {
  my ($class, %config) = @_;
  my $self = {namespace=>'resque', 
    %config};
  bless $self, $class;
}

# Returns the active Redis connection
sub redis {
  my $self = shift;
  unless ($self->{redis}) {
    my $server = $self->{config}{server}||'127.0.0.1:6379';
    $self->{redis} = Redis->new(server=>$server) or die "Can not connect to Redis: $server";
  }
  $self->{redis};
}

# Convenience method to return a new job on this resque system
sub create_job {
  my ($self, $klass, @args) = @_;
  Resque::Job::create($self, $klass, @args);
}

# Convenience method to return a new worker on this resque system
sub new_worker {
  my ($self, %config) = @_;
  new Resque::Worker(resque=>$self, %config);
}

# Used to decide to inline processing (testing) or queue the request
sub inline {
  my ($self) = @_;
  $self->{config}{inline}||undef;
}

sub key {
  my ($self, @names) = @_;
  my $name = join(':', @names);
  warn "namespace! @names ... ".join(',',caller) if $self->{namespace} eq 'resque'; ### TESTING
  "$self->{namespace}:$name";
}

sub keys {
  my ($self, @names) = @_;
  $self->redis->keys($self->key(@names,"*"));
}

sub queue {
  my ($self, $name) = @_;
  $self->key("queue", $name);
}

# Adds a job item to the named queue
sub push_queue {
  my ($self, $queue, $message) = @_;
  $self->watch_queue($queue);
  $self->redis->rpush($self->queue($queue), $message);
}

# Removes a job item to the named queue and returns it
sub pop_queue {
  my ($self, $queue) = @_;
  $self->redis->lpop($self->queue($queue));
}

sub available_logs {
  my ($self) = @_;
  $self->keys('logs');
}

sub push_log_rec {
  my ($self, $file, $rec) = @_;
  $self->redis->rpush($self->key('log', $file), $rec);
}

sub pop_log_rec {
  my ($self, $log_key) = @_;
  $self->redis->lpop($log_key);
}

# Returns the jobs from the queue
sub peek {
  my ($self, $queue, $start, $count) = @_;
  $count ||= 1;
  my @jobs=();
  if ($count == 1) {
    push @jobs,$self->redis->lindex($self->queue($queue), $start||0);
  }
  else {
    @jobs = $self->redis->lrange($self->queue($queue), $start||0, $count||1);
  }
  #@jobs = map { decode_json($_) } @jobs;
  @jobs;
}

sub queues {
  my ($self) = @_;
  $self->redis->smembers($self->key("queues"));
}

sub remove_queue {
  my ($self, $name) = @_;
  $self->redis->srem($self->key("queues"), $self->queue($name));
  delete $self->{watched_queues}{$name};
  $self->redis->del($self->queue($name));
}

# Private: Keep track of queues we've created
sub watch_queue {
  my ($self, $name) = @_;
  $self->{watched_queues} ||= {};
  $self->{watched_queues}{$name} = 1;
  $self->redis->sadd($self->key("queues"), $self->queue($name));
}

# Drops the queue and all jobs in it.
sub drop_queue {
  my ($self, $queue) = @_;
  my $name = $self->queue($queue);
  $self->{watched_queues} ||= {};
  delete $self->{watched_queues}{$name};
  $self->redis->srem($self->key("queues"), $self->queue($name));
  $self->redis->del($name);
}

# Drops all subordinate keys for the given key names
sub drop_keys {
  my ($self, @names) = @_;
  my @k = $self->keys();
  #$self->redis->del(@k); # Doesn't take multiple keys like the redis does
  foreach (@k) {
    #print "deleting key $_\n";
    $self->redis->del($_);
  }
}

# Seriously, will drop the entire queue data
sub drop_all {
  my ($self) = @_;
  $self->drop_keys();
}

# Enqueues a job in Resque-standard format
sub enqueue {
  my ($self, $obj, @args) = @_;
  my $queue = $obj->{queue_name} || $obj->queue_name() || ref($obj);
  $self->push_job($queue, {class=>ref($obj), args=>\@args});
}

# Destroys all jobs in the queue matching the arguments
sub dequeue {
  my ($self, $queue, @args) = @_;
  # This doesn't do as advertised
  my $job = $self->pop_job($queue);
  $job ? decode_json($job) : undef;
}

sub reserve {
  my ($self, $queue) = @_;
  $self->{reserve} = $self->dequeue($queue);
}

sub now {
  my ($sec,$min,$hour,$mday,$mon,$year,$wday, $yday,$isdst)=localtime(time);
  sprintf "%4d-%02d-%02d %02d:%02d:%02d", $year+1900,$mon+1,$mday,$hour,$min,$sec;
}

# Appends message to log file
sub logger {
  my ($self, $process, @msg) = @_;
  my $logrec = join("\t", now(), $process, @msg);
  return print "$logrec\n" if $self->{test};

  my $logfile = $self->{logfile} || $ENV{RESQUE_LOG} || "$process.log";
  open(my $lh, ">>", $logfile);
  return print $lh "$logrec\n" unless $lh;
  print $lh "$logrec\n";
  close $lh;
}

# Preloaded methods go here.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Resque - Perl extension for blah blah blah

=head1 SYNOPSIS

  use Resque;
  blah blah blah

=head1 DESCRIPTION

Stub documentation for Resque, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head2 EXPORT

None by default.



=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

Allen Fair, E<lt>allen@nonetE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Allen Fair

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.3 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Resque/Job.pm
package Resque::Job;

use 5.006000;
use strict;
use warnings;

use JSON::XS;
#use lib "$ENV{HOME}/src/Resque";
use Resque;
use Resque::Failure;

require Exporter;

our @ISA = qw(Exporter Resque);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Resque ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

my $to_s;

sub new {
  my ($class, $queue, $payload, %config) = @_;
  my $job = $payload ? decode_json($payload) : {};
  my $self = {queue=>$queue, payload=>$payload, job=>$job, %config};
  $self = $config{resque} ? {%{$config{resque}},%$self} : new Resque(%$self);
  bless $self, $class;
}

sub job_string {
  my ($klass, @args) = @_;
  encode_json({class=>$klass, args=>\@args});
}

sub to_s {
  my ($self) = @_;
  encode_json($self->{job});
}

# Creates a Job my placing it on a queue. Resque::Job::create(Class, *args..., {opt=>value,...})
sub create {
  my ($resque, $klass, @args) = @_;
  my $opt = ref($args[-1]) eq 'HASH' ? pop(@args) : {};
  return perform($klass, @args) if $resque->inline;
  $klass = ref($klass) if ref($klass);
  my $queue = $opt->{queue} || "$klass";
  my $job = new Resque::Job($queue, job_string($klass, @args), resque=>$resque, %$opt);
  $job->enqueue();
  $job;
}

sub enqueue {
  my ($self, $queue) = @_;
  $queue ||= $self->{queue};
  $self->push_queue($queue, $self->to_s);
}

# Removes all matching jobs from the queue, returns number of destroyed jobs
sub destroy {
  my ($resque, $queue, $klass, @args) = @_;
  my $destroyed=0;
  if (@args) {
    $destroyed = $resque->redis->lrem($resque->key('queue', $queue), 0, job_string($klass, @args));
  }
  else {
    my @payloads = $resque->redis->lrange($resque->key('queue', 0, -1));
    foreach my $p (@payloads) {
      my $job = decode_json($p);
      $destroyed += $resque->redis.lrem($resque->key('queue', 0, $p)) if $job->{class} eq $klass;
    }
  }
  $destroyed;
}

# Returns a job instance from the queue, or undef if none available
sub reserve {
  my ($resque, $queue) = @_;
  my $payload = $resque->pop_queue($queue);
  return undef unless $payload;
  new Resque::Job($queue, $payload, resque=>$resque);
}

sub args {
  @{$_[0]->{job}{args}||[]};
}

sub perform {
  my ($self) = @_;
  my $job_was_performed = 0;
  my $result;
  eval {
    no strict 'refs';
    $result = &{"$self->{job}{class}::perform"}($self->args);
    $job_was_performed = 1;
    use strict;
  };
  if ($@) { # Failure
    $self->fail($@);
  }
  $result;
}

sub fail {
  my ($self, $exception) = @_;
  print "job failed :-( $exception\n";
  Resque::Failure::create(payload=>$self->{payload}, exception=>$exception, 
    worker=>$self->{worker}, queue=>$self->{queue}, retries=>$self->{retries}||0);
}


1; # End of Package


lib/Resque/Worker.pm
package Resque::Worker;

use 5.006000;
use strict;
use warnings;

use Redis;
use JSON::XS;

require Exporter;

our @ISA = qw(Exporter Resque);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Resque ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

my $hostname;
my %workers=();
my $to_s;

sub new {
  my ($class, %config) = @_;
  my $self = {verbose=>0, queues=>$ENV{QUEUES}||'*', state=>'idle', %config};
  $self = $config{resque} ? {%{$config{resque}},%$self} : new Resque(%$self);
  bless $self, $class;
  $self->register_worker();
  $self;
}

# Removes a non-processing worker data
sub destroy {
  my ($self) = @_;
  $self->unregister_worker();
}

sub hostname {
  $hostname ||= `hostname`;
  chomp $hostname;
  $hostname;
}

sub to_s {
  my ($self) = @_;
  $to_s ||= join(":", hostname(), $$, join(',',$self->{queues}));
}

sub register_worker {
  my ($self) = @_;
  $self->prune_dead_workers();
  $workers{$self->to_s} = $self;
  $self->redis->sadd($self->key('workers'), $self->to_s);
  $self->redis->set($self->key('processed', $self->to_s), '0');
  $self->redis->set($self->key('worker', $self->to_s, 'started'), $self->now());
}

sub unregister_worker {
  my ($self) = @_;
  delete $workers{$self->to_s};
  $self->drop_keys('worker', $self->to_s);
  $self->redis->del($self->key('processed', $self->to_s));
  $self->redis->srem($self->key('workers'), $self->to_s);
}

# Returns the next job from the subscribed queues
sub reserve {
  my ($self) = @_;
  my @q = $self->{queues} eq '*' ? sort $self->queues() : split(',',$self->{queues});
  foreach (@q) {
    $self->{job} = Resque::Job::reserve($self, $_);
    return $self->{job} if $self->{job};
  }
}

# Returns an array of all known workers. Call with resque base class or subclass
sub workers {
  my ($resque) = @_;
  $resque->redis->smembers($resque->key("workers"));
}

sub prune_dead_workers {
  my ($self) = @_;
  my @rworkers = $self->workers();
  foreach my $w (@rworkers) {
    my ($pid) = $w =~ /:(\d+):/;
    my $wps = `ps -o pid,state -p $pid`;
    $self->redis->srem($self->key('workers'), $w) unless $wps;
    $self->redis->del($self->key('worker', $w, 'started'));
    $self->redis->del($self->key('processed', $w));
  }
}

# Increments the named status count for the worker and system
sub status {
  my ($self, $name) = @_;
  $self->redis->incrby($self->key('stat', $name), 1);
  $self->redis->incrby($self->key('stat', $self->to_s, $name), 1);
}

# Signals: QUIT=exit, TERM=kill child & exit, USR1=kill child & continue, USR2=Pause, CONT=Unpause
sub register_signal_handlers {
  my ($self) = @_;
  $self->{paused} = 0;
  $self->{shutdown} = 0;
  $self->{died} = 0;
  $SIG{INT}  = sub { $self->shutdown_now("Interrupted") };
  $SIG{TERM} = sub { $self->shutdown_now("Caught TERM") };
  $SIG{QUIT} = sub { $self->shutdown("Caught QUIT") };
  $SIG{USR1} = sub { $self->kill_child("Caught USR1") };
  $SIG{USR2} = sub { $self->pause_processing("Caught USR2") };
  $SIG{CONT} = sub { $self->unpause_processing("Caught CONT") };
#   $SIG{__DIE__} = sub { 
#     return if $self->{died};
#     if ($self->{child} && $self->{child} == $$) { # I'm a child
#       ++$self->{died};
#       $self->logger("Child died...",  @_);
#       exit 255;
#     }
#     else { # Master died. WTF?
#       $self->logger("Master died...",  @_);
#       exit 255;
#     }
#   };
}

sub unregister_signal_handlers {
  my ($self) = @_;
  delete $SIG{INT};
  delete $SIG{TERM};
  delete $SIG{QUIT};
  delete $SIG{USR1};
  delete $SIG{USR2};
  delete $SIG{CONT};
 #delete $SIG{__DIE__};
}

sub work {
  my ($self) = @_;
  $self->redis->sadd($self->key('workers'), $self->to_s);
  $self->redis->set($self->key('stat', 'processed', $self->to_s), "0");
  $self->register_signal_handlers();

  while(!$self->shutdown_requested) {
    if (!$self->paused && $self->reserve()) {
      my $rc;
      $self->working_on();
      $self->{child} = fork();
      $self->shutdown("Fork failed!") unless defined $self->{child};

      if ($self->{child} == 0) { # Child!
        srand;
        # Send job to a configired callback, or let the job do it.
        if ($self->{callback}) {
          $rc = &{$self->{callback}}($self->{job}->args());
        } else {
          $rc = $self->{job}->perform();
        }
        $self->status('processed');
        exit $rc;
      } 
      else { # Me, $self->{child} is the child pid
        my $cpid = wait();
        $self->logger("Lost my child! Sad Parent!") if $cpid == -1;
      }
      $self->{child} = 0;
      $self->done_working();
    } 
    elsif ($self->{shutdown_on_empty}) {
      $self->shutdown();
    } 
    else {
      sleep($self->{interval} || 5);
    }
  }
}

# Log Process Worker: reads base:logs:logname and writes lines to given logname
sub log_processor {
  my ($self) = @_;
  $self->redis->sadd($self->key('workers'), $self->to_s);
  $self->redis->set($self->key('stat', 'processed', $self->to_s), "0");
  $self->register_signal_handlers();
  
  while(!$self->shutdown_requested) {
    my @logs = $self->available_logs();
    if (!$self->paused && scalar(@logs)) {
      my $clog = '';
      my ($lh, $rec);
      foreach my $l (@logs) {
        my ($lname) = $l =~ /^.+?:log:(.+)/;
        while (!$self->paused && ($rec = $self->pop_log_rec($l))) {
          if ($clog ne $lname) {
            $clog = $lname;
            close $lh if $lh;
            open $lh, '>>', $lname;
          }
          print $lh "$rec\n" if $lh;
        }
        close $lh if $lh;
      }
    }
    elsif ($self->{shutdown_on_empty}) {
      $self->shutdown();
    } 
    else {
      sleep($self->{interval} || 5);
    }
  }
  $self->redis->srem($self->key('workers'), $self->to_s);
  $self->redis->del($self->key('stat', 'processed', $self->to_s));
}

# Updates base:worker:id to current job
sub working_on {
  my ($self, $job) = @_;
  $job ||= $self->{job};
  my $v = {queue=>$self->{queue}, run_at=>Resque::now(), payload=>$job->{payload}};
  $self->redis->set($self->key('worker',$self->to_s), encode_json($v));
}

sub done_working {
  my ($self) = @_;
  $self->redis->del($self->key('worker',$self->to_s));
}

# Call when job failed.
sub fail_job {
  my ($self, $exception) = @_;
  my $job ||= $self->{job};
  my $v = JSON::encode_json({queue=>$self->{queue}, run_at=>Resque::now(),
    payload=>$job, exception=>$exception, worker=>$self->to_s,
    failed_at=>Resque::now()});
  $self->redis->rpush("resque:failed", $v);
  $self->redis->incrby("resque:stat:failed:$self->{worker}", 1);
  $self->redis->incrby("resque:stat:failed", 1);
}

sub retry_queue {
  my ($self, $queue) = @_;
  while (my $obj = $self->dequeue_job("resque:queue:$self->{queue}")) {
    $self->queue_job($queue, $obj);
  }
}

# Finishes current job and halts
sub pause_processing {
  my ($self, @msg) = @_;
  $self->logger("Pausing...", @msg);
  $self->{paused} = time;
}

# Finishes current job and halts
sub unpause_processing {
  my $self = shift;
  $self->logger("Continuing...");
  $self->{paused} = 0;
}

# Finishes current job and halts
sub shutdown {
  my ($self, @msg) = @_;
  $self->logger("Exiting...", @msg);
  $self->{shutdown} = time;
}

# Cancels current job and halts
sub shutdown_now {
  my ($self, @msg) = @_;
  $self->kill_child(@msg);
  $self->shutdown(@msg);
}

# Finishes current job and halts
sub shutdown_requested {
  $_[0]->{shutdown} ? 1 : 0;
}

# Finishes current job and halts
sub paused {
  $_[0]->{paused} ? 1 : 0;
}

sub kill_child {
  my ($self, @msg) = @_;
  my $info = $self->child_process_info() or return 0;
  $self->logger("Killing child at $self->{child}", $info, @msg);
  `kill -9 $self->{child}`;
}

sub child_process_info {
  my $self = shift;
  return '' unless $self->{child};
  my $info = `ps -o pid,state -p $self->{child}`;
  return undef if $?; # Not found
  chomp $info;
  $info;
}

# Verbose logger
sub vlog {
  my ($self, @msg) = @_;
  $self->logger(@msg) if $self->{verbose};
}

# Appends message to log file
sub logger {
  my ($self, @msg) = @_;
  Resque::logger($self, $self->to_s(), @msg);
}


1; # End of Package
